### PR TITLE
ci: temp machines for scheduled killing experiment

### PR DESCRIPTION
**Pull request by garyverhaegen-da**
Based on our discussions last week, I am exploring ways to move us to permanent machines instead of preemptible ones. This should drastically reduce the number of "cancelled" jobs.

The end goal is to have:

1. An instance group (per OS) that defines the actual CI nodes; this would be pretty much the same as the existing ones, but with `preemptible` set to false.
2. A separate machine that, on a cron (say at 4AM UTC), destroys all the CI nodes.

The hope is that the group managers, which are set to maintain a 10 nodes, will then recreate the "missing" nodes using their normal starting procedure.

However, there are a lot of unknowns I would like to explore, and I need a playground for that. This is where this PR comes in. As it stands, it creates one "killer" machine and a temporary group manager. I will use these to experiment with the GCP API in various ways without interfering with the real CI nodes.

This experimentation will likely require multiple `terraform apply` with multiple different versions of the associated files, as well as connecting to the machines and running various commands directly from them. I will ensure all of that only affects the new machines created as part of this PR, and therefore believe we do not need to go through a separate round of approval for each change.

Once I have finished experimenting, I will create a new PR to clean up the temporary resources created with this one and hopefully set up a more permanent solution.

### CODE REVIEW

File: infra/TEMP_KILLABLE_vsts_agent_linux.tf [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+  service_account {
+    email  = "log-writer@da-dev-gcp-daml-language.iam.gserviceaccount.com"
+    scopes = ["cloud-platform"]
+  }
+
+  scheduling {
+    automatic_restart   = false
+    on_host_maintenance = "TERMINATE"
+    preemptible         = true
```
Comment (cocreature) [inline infra/TEMP_KILLABLE_vsts_agent_linux.tf]: I thought the end goal is to have these not be preemptible. I guess for testing it just doesn’t matter?

Comment (garyverhaegen-da) [reply infra/TEMP_KILLABLE_vsts_agent_linux.tf]: Preemptible means 20% of the price, so there's no reason not to in this case.

**Comment by garyverhaegen-da** [comment]
> I'm also fine with playing multiple rounds of terraform apply and documenting them on this PR.

Alright then... I'll try to figure ut how to do that.

**Comment by garyverhaegen-da** [comment]
New plan then: I will not merge this PR until I'm done experimenting, and every time I want to do a `terraform apply` I'll first make a commit and push it to this branch. @hurryabit Do I need to then wait for approval on each of those?

**Comment by garyverhaegen-da** [comment]
Applying d72a2743688f067bbb83782b99278698adce8bae failed because I forgot to specify a zone for the killer instance. The template and group manager were successfully created.

No other resource was touched, as expected.

**Comment by garyverhaegen-da** [comment]
Next step: I'd like to try applying d12392871936f5cb236e20731ee8bd61035572de, which hopefully should resolve the "missing zone" issue.

**Comment by hurryabit** [comment]
> New plan then: I will not merge this PR until I'm done experimenting, and every time I want to do a `terraform apply` I'll first make a commit and push it to this branch. @hurryabit Do I need to then wait for approval on each of those?

No, you don't need to wait.

**Comment by garyverhaegen-da** [comment]
Successfully applied d12392871936f5cb236e20731ee8bd61035572de, creating the instance as expected (and not touching anything else).

**Comment by garyverhaegen-da** [comment]
Tried to apply fcd136817ba6adec19565f891ac7fd0c20974ac8, adding total control over all Compute Instances to the killer machine. What it really needs is authorization to list them and kill them, but I could not find anything more granular.

Apply didn't get through, though, because apparently changing the scopes requires a machine restart and Terraform doesn't like doing that without explicit permission.

**Comment by garyverhaegen-da** [comment]
Applied 378c7606a1039c657e6ee937aff87b8aef2a337e, recreating the killer machine with (hopefully) the right set of permissions.

**Comment by garyverhaegen-da** [comment]
Trying to list instances from the killer machine results in missing permission "compute.zones.list". Further digging in the documentation reveals that scopes are hard to understand and quite useless; the [recommended "best practice"](https://cloud.google.com/compute/docs/access/service-accounts#accesscopesiam) is to set the scopes to "everything" (i.e. `cloud-platform`) and then use IAM permissions on the service account. But somehow they chose to not make that the default.

Tried to apply 609457fde7f823e36387229999999dd6d53f15d0 to add a service account. It failed because I apparently do not have the permission to create a custom IAM role. It did create the service account though.

**Comment by garyverhaegen-da** [comment]
Now that Ed has given me more access rights, I have applied 609457fde7f823e36387229999999dd6d53f15d0 again, this time successfully creating the new role.

**Comment by garyverhaegen-da** [comment]
Tried to apply 3b2b189a9bdf7bd6e046aee43dc58d14f98af892, which adds a link between the IAM role and the service account.

Failed with two errors:
1. google_project_iam_custom_role.periodic-killer: Custom project role projects/da-dev-gcp-daml-language/roles/killCiNodesEveryNight already exists and must be imported
2. google_project_iam_member.periodic-killer: Error applying IAM policy for project "da-dev-gcp-daml-language": Error setting IAM policy for project "da-dev-gcp-daml-language": googleapi: Error 403: The caller does not have permission, forbidden

The first one is due to the GCP API being a bit difficult: because the (Terraform) name of the resource has changed, Terraform thinks it needs to destroy the old resource, which is no longer present, and create a new one. So it sends the command to delete, gets an OK from GCP, then sends a command to create a new one with the same ID, and gets an error from GCP because role deletion on GCP is "eventually consistent", and apparently Terraform goes too fast.

The second error seems to indicate I still do not have enough access rights. I'd really like to understand why GCP decided it would make sense to have an access level that can create new roles but not actually use them. (Or why it would ever make sense to have the ability to use the default roles, which include, like, "can do everything", but not the ability to create and assign custom roles with fewer access rights.)

**Comment by garyverhaegen-da** [comment]
Applying 3b2b189a9bdf7bd6e046aee43dc58d14f98af892 again now that time has passed and Ed has further expanded my access rights yields an interesting new error: while the custom role has now been created successfully, the member rule has failed because Terraform tried to apply it _before_ it created the role (which GCP interpreted as trying to add to a deleted role, because for some reason it keeps track of those):
> google_project_iam_member.periodic-killer: Error applying IAM policy for project "da-dev-gcp-daml-language": Error setting IAM policy for project "da-dev-gcp-daml-language": googleapi: Error 400: Cannot add new grants to deleted role (projects/da-dev-gcp-daml-language/roles/killCiNodesEveryNight)., badRequest

**Comment by garyverhaegen-da** [comment]
Applying 3b2b189a9bdf7bd6e046aee43dc58d14f98af892 yet again now that a few minutes have passed finally succeeded. Yay for eventual consistency. 😞

**Comment by garyverhaegen-da** [comment]
Manually connecting to the killer machine and running `gcloud compute instances list` now results in
```
ERROR: (gcloud.compute.instances.list) Some requests did not succeed:
 - Required 'compute.instances.list' permission for 'projects/da-dev-gcp-daml-language'
```

**Comment by garyverhaegen-da** [comment]
Applied 9203f9fb216fad6060fbe5426b371f4054fb546f successfully to add the missing permission.

**Comment by garyverhaegen-da** [comment]
`gcloud compute instances list` now succeeds. Trying the next step, `gcloud -q compute instances delete temp-killable-rtqn --zone=us-east4-c`, predictably fails for lack of permissions:
```
$ gcloud -q compute instances delete temp-killable-rtqn --zone=us-east4-c
ERROR: (gcloud.compute.instances.delete) Could not fetch resource:
 - Required 'compute.instances.delete' permission for 'projects/da-dev-gcp-daml-language/zones/us-east4-c/instances/temp-killable-rtqn'
```

**Comment by garyverhaegen-da** [comment]
Successfully applied 37b9e1259eceb68191d67e73fba1ce881f1a8066, adding the `compute.instances.delete` permission to the killer machine.

**Comment by garyverhaegen-da** [comment]
Successfully ran `gcloud -q compute instances delete temp-killable-rtqn --zone=us-east4-c`, which should be all the permissions this machine needs. Also confirmed through visual inspection in the web console that the group manager did create a replacement instance, which means GCP is behaving as expected in that regard.

**Comment by garyverhaegen-da** [comment]
Applied 080e37d8960f30e6e9638f96589be87798bdad54, which should set up a cron that kills all the (`temp-killable-*`) machines every minute.

**Comment by garyverhaegen-da** [comment]
Applied 30bf7ab7a1524ca624ff4e599a67b7a2dea21ab5, which tweaks the init script so it (hopefully) inserts dollars into the crontab script rather than trying to resolve them as part of the init script.

**Comment by garyverhaegen-da** [comment]
Escaping now works but apparently crontab entries run with a more limited PATH. Applied bcacb05eaad32a00a818fe5008736d1258fea5f3 to try to fix that.

**Comment by garyverhaegen-da** [comment]
Everything now seems to be working fine. I've applied 6a5f3a8c88f76eb84d7f530af7864d48e0cad0f9 with final tweaks (reboot every 30 minutes rather than every minute, and use a smaller machine for the killer).

I'll merge this in as it represents the current state of things, and monitor over the weekend. If everything seems fine I'll open a PR on Monday to turn this on for the real CI nodes.

**Pull request merged**